### PR TITLE
Add Identity-Tasklist integration

### PR DESCRIPTION
**Pull request by Zelldon**
Blocked by #268 (contains currently lot of commits from 268, will be rebased after merging #268)

Almost similar to #268 

> 
> - [x] Configure Tasklist and Identity to communicate with each other
>   - [x] Add env vars to identity
>   - [x] Add env vars to tasklist
> - [x] Add integration test for tasklist login (use a session based login here)
> - [x] Auto generate a secret for Operate-Identity ( can be also overwritten by user and an existing secret)
> - [x] Adjust docs, values file
> - [x] Make it possible to disable this configuration
> - [x] Add notes in Helm release notes file
> - [x] Add template tests
> 

closes https://github.com/camunda/camunda-platform-helm/issues/270

### CODE REVIEW

**Comment by npepinpe** [comment]
Does this contain the commits from the Operate PR? It kind of looks like it does, and it would explain the size of the PR :smile:

File: charts/camunda-platform/templates/NOTES.txt
Comment (npepinpe) [inline charts/camunda-platform/templates/NOTES.txt:79]: :wrench: Same as the other PR, wording is kind of funky here, I would remove the `Per default` and only mention instead to change it if they didn't want it.

File: charts/camunda-platform/test/tasklist/deployment_test.go
Comment (npepinpe) [inline charts/camunda-platform/test/tasklist/deployment_test.go:203]:
```suggestion
func (s *deploymentTemplateTest) TestContainerShouldSetTasklistIdentitySecret() {
```

File: charts/camunda-platform/test/tasklist/deployment_test.go
Comment (npepinpe) [inline charts/camunda-platform/test/tasklist/deployment_test.go:175]:
```suggestion
func (s *deploymentTemplateTest) TestContainerShouldDisableTasklistIntegration() {
```

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1: 

@dlavrenuek Same as the other PR, I can't judge fully if the config is right, but it starts so good enough for me. Please check more in details, as you can probably tell better.

Some notes:

- :x: I just realized we have to port-forward keycloak. That's less than ideal :smile: Is there any way it can just work without that? Because how would this look like in production? They'd need to run a port-forward all the time? Yikes. Did I miss some other way of doing it? If they have an ingress for Keycloak then it will work fine? I find that a bit strange, I would expect all authentication requests to go through Identity, and Keycloak to be fully hidden.
- I think versions can be updated to 8.0.0

**Comment by Zelldon** [comment]
Hey @npepinpe thanks for your review :+1: 

Regarding your comments/questions:



> I just realized we have to port-forward keycloak. That's less than ideal smile Is there any way it can just work without that? Because how would this look like in production? They'd need to run a port-forward all the time? Yikes. Did I miss some other way of doing it? If they have an ingress for Keycloak then it will work fine? I find that a bit strange, I would expect all authentication requests to go through Identity, and Keycloak to be fully hidden.

YES. This is what I raised multiple times now as well, but it was considered as the way to go for now. Tbh I saw also currently no other better way. The login is done via redirects to Keycloak and back. As alternative the user can deploy an ingress controller and enable the ingress for the services then it might be nicer to use (instead of port-forwards).

Slack comment about this: https://camunda.slack.com/archives/C02UT3T8DQR/p1648477086721989
Github comment: https://github.com/camunda/camunda-platform-helm/issues/127#issuecomment-1078722653

> I think versions can be updated to 8.0.0

I will do it after merging the PR's. 


> Does this contain the commits from the Operate PR? It kind of looks like it does, and it would explain the size of the PR smile

Yes  sorry. I actually thought I mentioned this in the description.

File: charts/camunda-platform/test/identity/golden/deployment.golden.yaml
Comment (dlavrenuek) [inline charts/camunda-platform/test/identity/golden/deployment.golden.yaml:65]: The port `8080` is used for all three services. Identity uses this configuration to set the frontend url (f.e. login callback url for OAuth), so if there will be another routing to another port/domains, then `IDENTITY_URL` has to be set

Comment (Zelldon) [reply charts/camunda-platform/test/identity/golden/deployment.golden.yaml:65]: Sorry not 100% sure whether I can follow. Do you think it might be better if we just configure for all different ports and mention this in the notes to port-forward also to different ports? Then we adjust here the config as well?

Like:

8080 always for identity
8081 for operate
8082 for tasklist

etc.

**Review by oleschoenburg — COMMENTED** [review]
I had first look and the only thing that stood out to me is how complex the templating for the identity config is becoming. It'd probably require too much refactoring but I'm wondering if it'd be possible/preferable to just allow free form yaml in the values.yml and generate a configmap to be used by identity.

File: charts/camunda-platform/charts/tasklist/templates/deployment.yaml
Comment (Zelldon) [reply charts/camunda-platform/charts/tasklist/templates/deployment.yaml]: I refactored the values file a bit thanks for the feedback :)

**Comment by oleschoenburg** [comment]
> Do you have any specific proposal otherwise I would say we can iterate over it later

@Zelldon nothing more specific than what I already wrote down. I agree that getting this working first and then iterating later is fine :+1:

**Comment by Zelldon** [comment]
In order to make the starting experience easier configured the redirectUrls to different ports and adjusted the release helm notes such that it makes it easier to play around with it. We can just copy the port-forwards and get access to all web applications.


```
NAME: camunda-platform-test
LAST DEPLOYED: Tue Apr 12 12:09:46 2022
NAMESPACE: zell-helm-test
STATUS: deployed
REVISION: 1
NOTES:
# (camunda-platform - 0.0.29)
   ___   _   __  __ _   _ _  _ ___   _     ___ _      _ _____ ___ ___  ___ __  __ 
  / __| /_\ |  \/  | | | | \| |   \ /_\   | _ \ |    /_\_   _| __/ _ \| _ \  \/  |
 | (__ / _ \| |\/| | |_| | .` | |) / _ \  |  _/ |__ / _ \| | | _| (_) |   / |\/| |
  \___/_/ \_\_|  |_|\___/|_|\_|___/_/ \_\ |_| |____/_/ \_\_| |_| \___/|_|_\_|  |_|
                                                                                 

## Installed Services:

- Zeebe:
  - Docker Image used for Zeebe: camunda/zeebe:8.0.0-rc1
  - Zeebe Cluster Name: "camunda-platform-test-zeebe"
  - Prometheus ServiceMonitor Enabled: false
- Operate:
  - Enabled: true
  - Docker Image used for Operate: camunda/operate:8.0.0-rc1
- Tasklist:
  - Enabled: true
  - Docker Image used for Tasklist: camunda/tasklist:8.0.0-rc1
- Identity:
  - Enabled: true
  - Docker Image used for Identity: camunda/identity:8.0.0-rc1
  - Keycloak: bitnami/keycloak:16.1.1-debian-10-r52
- Elasticsearch:
  - Enabled: true
  - ElasticSearch URL: http://elasticsearch-master:9200

### Zeebe

The Cluster itself is not exposed as a service that means that you can use `kubectl port-forward` to access the Zeebe cluster from outside Kubernetes:

> kubectl port-forward svc/camunda-platform-test-zeebe-gateway 26500:26500 -n zell-helm-test

Now you can connect your workers and clients to `localhost:26500`

### Connecting to Web apps

As part of the Helm charts an ingress definition can be deployed, but you require to have an Ingress Controller for that Ingress to be Exposed.
In order to deploy the ingress manifest, set `<service>.ingress.enabled` to `true`. Example: `operate.ingress.enabled=true`

If you don't have an ingress controller you can use `kubectl port-forward` to access the deployed web application from outside the cluster:

Identity: kubectl port-forward svc/camunda-platform-test-identity 8080:80
Operate:  kubectl port-forward svc/camunda-platform-test-operate 8081:80
Tasklist: kubectl port-forward svc/camunda-platform-test-tasklist 8082:80

If you want to use different ports for the services, please adjust the related configs in the values file since these ports are used as redirect URL for Keycloak.

The authentication via Identity/Keycloak is enabled. In order to login into one of the services please port-forward to Keycloak
as well, otherwise a login will not be possible. Make sure you use `18080` as port.

> kubectl port-forward svc/camunda-platform-tes 18080:80

Now you can point your browser to one of the service's login page. Example: http://localhost:8081 for Operate.

Default user and password: "demo/demo"
```

I can now easily port-forward and access them:

console
webapp

File: charts/camunda-platform/test/identity/golden/deployment.golden.yaml
Comment (Zelldon) [reply charts/camunda-platform/test/identity/golden/deployment.golden.yaml]: thanks for your input @dlavrenuek ! I reworked it a bit see here https://github.com/camunda/camunda-platform-helm/pull/271#issuecomment-1096514410

**Comment by Zelldon** [comment]
If no one objects until 1 pm @npepinpe @oleschoenburg @dlavrenuek I will merge it and do a new release (after I updated to cam 8 and do some other smaller pieces)

File: charts/camunda-platform/charts/identity/templates/deployment.yaml
Comment (dlavrenuek) [reply charts/camunda-platform/charts/identity/templates/deployment.yaml:59]: Sounds good 👍

**Pull request merged**
